README
--------------------------------------------------------------------------------
     fsmap:  Display a file hierarchy in filesystem markup language (FSML)
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
description
--------------------------------------------------------------------------------
NAME
  fsmap - display a file hierarchy in filesystem markup language (FSML)

SYNOPSIS
  fsmap [path]

DESCRIPTION
  The fsmap utility recursively descends the directory tree for path and
  prints a filesystem markup language (FSML) representation of the traversed
  file hierarchy to standard out.

  If a regular file with a supported file extension is visited fsmap tries to
  extract file-specific metadata. Currently 215 file extension are supported
  via exiftool(1). PDF content is additionally extracted using pdftohtml(1).

EXAMPLE
 Basic usage:

  $ tree ~/Demo/Images/Database/
  /home/holu/Demo/Images/Database/
  └── XML
      └── BaseX.png

      1 directory, 1 file

  $ python fsmap.py ~/Demo/Images/Database/
  <?xml version="1.0"?>
  <fsml name="/home/holu/Demo/Images/Database" 
        st_ctime="1310499391" st_mtime="1310499391" st_nlink="3"
        st_gid="1000" st_dev="2049" st_size="4096" st_mode="040755"
        st_uid="1000" st_atime="1310499416">
    <dir name="XML"
        st_ctime="1310499403" st_mtime="1310499403" st_nlink="2"
        st_gid="1000" st_dev="2049" st_size="4096" st_mode="040755"
        st_uid="1000" st_atime="1310499416">
      <file name="BaseX.png" suffix=".png"
        st_ctime="1310499403" st_mtime="1310499403" st_nlink="1"
        st_gid="1000" st_dev="2049" st_size="76251" st_mode="0100644"
        st_uid="1000" st_atime="1310499416">
        <meta xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#r"
          xmlns:et='http://ns.exiftool.ca/1.0/' et:toolkit='Image::ExifTool 8.60'
          xmlns:PNG='http://ns.exiftool.ca/PNG/PNG/1.0/'
          xmlns:Composite='http://ns.exiftool.ca/Composite/1.0/'>
         <PNG:ImageWidth>1410</PNG:ImageWidth>
         <PNG:ImageHeight>1033</PNG:ImageHeight>
         <PNG:BitDepth>8</PNG:BitDepth>
         <PNG:ColorType>RGB with Alpha</PNG:ColorType>
         <PNG:Compression>Deflate/Inflate</PNG:Compression>
         <PNG:Filter>Adaptive</PNG:Filter>
         <PNG:Interlace>Noninterlaced</PNG:Interlace>
         <PNG:CreationTime>Mi 19 Jul 2006 09:40:17 +0100</PNG:CreationTime>
         <PNG:ModifyDate>2007:08:28 16:23:06</PNG:ModifyDate>
         <PNG:PixelsPerUnitX>23622</PNG:PixelsPerUnitX>
         <PNG:PixelsPerUnitY>23622</PNG:PixelsPerUnitY>
         <PNG:PixelUnits>Meters</PNG:PixelUnits>
         <PNG:Gamma>2.2</PNG:Gamma>
         <Composite:ImageSize>1410x1033</Composite:ImageSize>
        </meta>
      </file>
    </dir>
  </fsml>

HISTORY
  A first version of the filesystem markup language appeared as part of the
  Minerva (http://www.minervahome.net/) project by Steven Goodwin.
 
AUTHOR
  Copyright (c) 2011 Alexander Holupirek <alex@holupirek.de>

  fsmap is free software; you can use it under the terms of the ISC license.

SEE ALSO
  exiftool(1), pdftohtml(1)

--------------------------------------------------------------------------------
credits
--------------------------------------------------------------------------------
  * Phil Harvey
    ExifTool: Read, write and edit meta information in a wide variety of files
    http://www.sno.phy.queensu.ca/~phil/exiftool
  * Gueorgui Ovtcharov, Rainer Dorsch, and the poppler team for pdftohtml
    http://poppler.freedesktop.org

  Thank you for your great open source tools!

--------------------------------------------------------------------------------
license
--------------------------------------------------------------------------------
If not explicitly stated otherwise all fsmap(1) code is released under the
ISC license, i.e., the following terms apply:

/*
 * Copyright (c) 2011 Alexander Holupirek <alex@holupirek.de>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */


changelog
Version 0.02 (July 18, 2011): 

    fsmap(1) exiftool version:
    * [ADD] hook for metadata extractors (aka transducers) plug-ins
    * [ADD] transducer for 215 file extensions via fabulous exiftool(1)
            extracts metadata from 
            .3fr .3g2 .3gp .3gp2 .3gpp .acfm .acr .afm .ai .aif .aifc .aiff .ait
            .amfm .ape .arw .asf .avi .bmp .btf .ciff .cos .cr2 .crw .cs1 .dc3
            .dcm .dcp .dcr .dfont .dib .dic .dicm .divx .djv .djvu .dll .dng
            .doc .docm .docx .dot .dotm .dotx .dv .dvb .dylib .eip .eps .eps2
            .eps3 .epsf .erf .exe .exif .f4a .f4b .f4p .f4v .fla .flac .flv .fpx
            .gif .gz .gzip .hdp .htm .html .icc .icm .iiq .ind .indd .indt .itc
            .jng .jp2 .jpeg .jpg .jpm .jpx .k25 .kdc .key .kth .lnk .m2t .m2ts
            .m2v .m4a .m4b .m4p .m4v .mef .mie .mif .miff .mka .mks .mkv .mng
            .mos .mov .mp3 .mp4 .mpc .mpeg .mpg .mpo .mqv .mrw .mts .mxf .nef
            .newer .nmbtemplate .nrw .numbers .odp .ods .odt .ogg .orf .otf
            .pages .pbm .pct .pdf .pef .pfa .pfb .pfm .pgf .pgm .pict .pmp .png
            .pot .potm .potx .ppm .pps .ppsm .ppsx .ppt .pptm .pptx .ps .ps2
            .ps3 .psb .psd .psp .pspframe .pspimage .pspshape .psptube .qif .qt
            .qti .qtif .ra .raf .ram .rar .raw .rif .riff .rm .rmvb .rpm .rsrc
            .rtf .rv .rw2 .rwl .rwz .so .sr2 .srf .srw .svg .swf .thm .thmx .tif
            .tiff .ts .ttc .ttf .tub .vob .vrd .vsd .wav .wdp .webm .webp .wma
            .wmv .x3f .xcf .xhtml .xla .xlam .xls .xlsb .xlsm .xlsx .xlt .xltm
            .xltx .xmp .zip

Version 0.01 (July 9, 2011): 

    fsmap(1) initial version:
    * displays file hierarchy in FSML
    * includes stat(2) information as XML attributes


fsmap.py
#!/usr/bin/env python
'''
fsmap(1) - display a file hierarchy in FSML

Copyright (c) 2011 Alexander Holupirek <alex@holupirek.de>

ISC license
'''
import sys
import os
import transducers

# --- xml formatting and output ------------------------------------------------
def out(str_):
    '''Print string on standard output stream.'''
    sys.stdout.write(str_)

def outln(str_):
    '''Print string and a newline on standard output stream.'''
    out(str_)
    sys.stdout.write('\n')

def indent(str_, width):
    '''Pad string on the left with 2 * width spaces.'''
    ind = "".rjust(2 * width)
    return ind + str_

def escape_xml(string):
    '''Escape characters not allowed in attribute values and tag names'''
    esc = ""
    for char in string:
        if char == '&':
            esc += '&amp;'
        elif char == '<':
            esc += '&lt;'
        elif char == '>':
            esc += '&gt;'
        elif char == '"':
            esc += '&quot;'
        else:
            esc += char
    return esc

def start_tag(tagname, atts=None):
    '''Construct a start tag string.
    tagname: string
    atts   : dict
    '''
    attributes = ""
    if atts:
        for name in atts:
            attributes += (" " + name + "=\"" + str(atts[name]) + "\"")
    return "<{0}{1}>".format(tagname, attributes)

def end_tag(tagname):
    '''Construct end tag string.'''
    return "</{0}>".format(tagname)

def empty_element(tagname, atts):
    '''Construct an empty element string.
    tagname: string
    atts   : dict'''
    str_ = start_tag(tagname, atts)
    return str_[:-1] + "/>"

def add_stat_attributes(path, atts):
    '''Insert stat(2) info for path into atts dict to build XML attributes.
    path: string, absolute path to file
    atts: dictionary to insert'''
    try:
        stat_ = os.lstat(path)
        atts['st_mode']  = oct(stat_.st_mode)
        atts['st_dev']   = str(stat_.st_dev)
        atts['st_nlink'] = str(stat_.st_nlink)
        atts['st_uid']   = str(stat_.st_uid)
        atts['st_gid']   = str(stat_.st_gid)
        atts['st_size']  = str(stat_.st_size)
        atts['st_atime'] = str(int(stat_.st_atime))
        atts['st_mtime'] = str(int(stat_.st_mtime))
        atts['st_ctime'] = str(int(stat_.st_ctime))
    except OSError as err:
        sys.stderr.write('(add_stat_attributes): ' + str(err) + '\n')

def get_default_atts(path):
    '''Create dict with name and stat(2) info for path (to build XML attributes)
    path  : string, absolute path to file
    return: atts dict'''
    value = escape_xml(os.path.basename(path))
    atts = { 'name' : value }
    add_stat_attributes(path, atts)
    return atts

# --- file traversal events ----------------------------------------------------
def pre_traversal(path):
    '''Before file hierarchy traversal begins.
    path : string, absolute path to file'''
    outln('<?xml version="1.0"?>')
    if (not os.path.isdir(path)):
        path = os.path.dirname(path)
    atts = { 'name' : escape_xml(path) }
    add_stat_attributes(path, atts)
    outln(start_tag("fsml", atts))

def post_traversal():
    '''After file hierarchy traversal.'''
    outln(end_tag("fsml"))

def visit_enter_directory(path, depth):
    '''Enter directory during file hierarchy traversal.
    path : string, absolute path to directory'''
    atts = get_default_atts(path)
    outln(indent(start_tag("dir", atts), depth))

def visit_leave_directory(depth):
    '''Leave directory during file hierarchy traversal.'''
    outln(indent(end_tag("dir"), depth))

def visit_empty_directory(path, depth):
    '''Empty directory is encountered during file hierarchy traversal.'''
    atts = get_default_atts(path)
    outln(indent(empty_element("dir", atts), depth))

def visit_link(path, depth):
    '''Symbolic or hard link is encountered during file hierarchy traversal.'''
    atts = get_default_atts(path)
    outln(indent(empty_element("link", atts), depth))

def visit_file(path, depth=0, xdcr_map=None):
    '''Regular file is encountered during file hierarchy traversal.
    path    : string, absolute path to regular file
    depth   : depth of file in file hierarchy
    xdcr_map: dict{ suffix : extract callback function } metadata extractors'''
    atts = get_default_atts(path)
    suffix = os.path.splitext(path)[1]
    atts['suffix'] = str.lower(suffix)
    if xdcr_map:
        if suffix in xdcr_map:
            outln(indent(start_tag("file", atts), depth))
            metadata = xdcr_map[suffix](path, suffix) # call xdcr extract method
            if metadata:
                for xml in metadata:
                    out(indent(xml, depth + 1))
            outln(indent(end_tag("file"), depth))
            return
    outln(indent(empty_element("file", atts), depth))

def descend(dpath, depth=0, xdcr_map=None):
    '''Recursively descend into file hierachry.
    path    :  pathname to directory
    depth   : current depth in file hierarchy
    xdcr_map: metadata extractors (pass through to visit_file)'''
    try:
        dents = os.listdir(dpath)
        if depth != 0:
            if not dents:
                visit_empty_directory(dpath, depth)
                return
            visit_enter_directory(dpath, depth)
        for file_ in dents:
            path = os.path.join(dpath, file_)
            if os.path.islink(path):
                visit_link(path, depth + 1)
            elif os.path.isdir(path):
                descend(path, depth + 1, xdcr_map)
            else:
                visit_file(path, depth + 1, xdcr_map)
        if depth != 0:
            visit_leave_directory(depth)

    except OSError as err:
        sys.stderr.write('(descend): ' + str(err) + '\n')

def walk(path):
    '''Enter file hierarchy traversal or print single file mapping.
    path    : path to file'''
    abspath = os.path.abspath(path)
    pre_traversal(abspath)
    xdcr_map = xdcr_init()
    level = 0
    if (os.path.isdir(abspath)):
        descend(abspath, level, xdcr_map)
    else:
        visit_file(abspath, level + 1, xdcr_map)
    post_traversal()

# --- transducer initialization ------------------------------------------------
def xdcr_init():
    '''Initialize metadata extractors/transducers & populate callback table.'''
    xdcr_map = {}
    exiftool = transducers.TransducerExifTool()
    (exiftool_suffixes, exiftool_extract_func) = exiftool.expose_transducer()
    for suffix in exiftool_suffixes:
        xdcr_map[suffix] = exiftool_extract_func
    xdcr_map[''] = transducers.extract_mail
    xdcr_map['.mbs'] = transducers.extract_mail
    return xdcr_map

if __name__ == '__main__':
    VERSION = .02
    if len(sys.argv) != 2:
        raise SystemExit(
            'fsmap(1) - display a file hierarchy in FSML (v%.2f)\n' % VERSION +
            'Usage: %s <path>' % sys.argv[0])
    walk(sys.argv[1])


FirstModule.py
from lxml import etree
import os

def main():
    tree_root = etree.Element("mails")
    path = "D:\\UNI\\SS11\\BP\\syawrtie.default" #search this path for mboxes
    for root, dirs, filenames in os.walk(path):
        for filename in filenames:
            if "." not in filename: #detects files without extension as mbox...
                mbox = root + "\\" + filename
                scan_mbox(mbox, tree_root)
    #write xml-tree to file
    fobj = open("mails.xml", "wb")
    fobj.write(etree.tostring(tree_root, pretty_print=True))
    fobj.flush()
    fobj.close()
    
def scan_mbox(mbox, root):
    fobj = open(mbox, "rb")                              #mbox file to read
    a = 0
    mbox = etree.SubElement(root, "mbox", path=mbox)
    for line in fobj:                                   #walk file line by line
        line.strip()
        part = line.split()                             #splits line at whitespaces
        if len(part) > 1:
            if part[0] == b'From' and part[1] == b'-':    #new Message begins
                msg = etree.SubElement(mbox, "msg")
                a += 1
            if part[0] == b'To:':
                text = ""
                for x in range(1, len(part)):
                    try:
                        text += part[x].decode('utf-8') + " "
                    except:
                        text += part[x].decode('cp1252') + " "
                etree.SubElement(msg, "to").text = text
            if part[0] == b'From:':
                text = ""
                for x in range(1, len(part)):
                    text += part[x].decode('cp1252') + " "
                etree.SubElement(msg, "from").text = text
            if part[0] == b'Subject:':
                text = ""
                for x in range(1, len(part)):
                    text += part[x].decode('cp1252') + " "
                etree.SubElement(msg, "title").text = text
            if part[0] == b'Date:':
                text = ""
                for x in range(1, len(part)):
                    text += part[x].decode('cp1252') + " "
                etree.SubElement(msg, "date").text = text
    print("\nmails found: ")  #just for debugging
    print(a)

main()

transducers.py
# vim: set fileencoding=utf-8 :
'''
transducers.py - metadata extractors (aka transducers) for various filetypes

Copyright (c) 2011 Alexander Holupirek <alex@holupirek.de>

ISC license
'''
import atexit
import sys
import subprocess
from subprocess import PIPE
from lxml import etree

def extract_mail(path, suffix):
    tree_root = etree.Element("meta")
    fobj = open(path, "rb")                              #mbox file to read
    if suffix == "":
        mbox = etree.SubElement(tree_root, "mbox")
    if suffix == ".mbs":
        mbox = etree.SubElement(tree_root, "maildir")
    for line in fobj:                                   #walk file line by line
        line.strip()
        part = line.split()                             #splits line at whitespaces
        if len(part) > 1:
            if suffix == "":
                if part[0] == b'From' and part[1] == b'-':    #new Message begins
                    msg = etree.SubElement(mbox, "msg")
            if suffix == ".mbs":
                if part[0] == b'From':                  #new Message begins
                    msg = etree.SubElement(mbox, "msg")
            if part[0] == b'To:':
                text = ""
                for x in range(1, len(part)):
                        text += part[x].decode('cp1252') + " "
                etree.SubElement(msg, "to").text = text
            if part[0] == b'From:':
                text = ""
                for x in range(1, len(part)):
                    text += part[x].decode('cp1252') + " "
                etree.SubElement(msg, "from").text = text
            if part[0] == b'Subject:':
                text = ""
                for x in range(1, len(part)):
                    text += part[x].decode('cp1252') + " "
                etree.SubElement(msg, "title").text = text
            if part[0] == b'Date:':
                text = ""
                for x in range(1, len(part)):
                    text += part[x].decode('cp1252') + " "
                etree.SubElement(msg, "date").text = text
    tree = etree.tostringlist(tree_root, pretty_print=True)
    return tree

#  Minimal (empty) result (4 lines) --> will not be returned
#  <content>
#  <page number="1" position="absolute" top="0" left="0" ...>
#  </page>
#  </content>
def extract_pdftohtml(path):
    '''Extract text from PDF file.
    path     : string. absolute path to regular (pdf) file.
    pagecount: int. number of pages in pdf file.
    return   : list. extracted text lines wrapped in <page> elements.'''
    try:
        pdftext = [ '  <content>\n' ]
        text = subprocess.check_output( ['pdftohtml', '-nodrm', '-xml'
            , '-enc', 'UTF-8', '-stdout', path])
        lines = str(text).splitlines(True)
        # strip first 4 (header) and last two lines (appendix)
        del lines[:4]
        del lines[-1:]
        has_text = False
        for line in lines:
            if line.lstrip().startswith("<text"):
                has_text = True
                # substitute common ligatures
                # www.alanwood.net/unicode/alphabetic_presentation_forms.html
                line = line.replace("ﬀ ", "ff")   # \uFB00
                line = line.replace("ﬁ", "fi")    # \uFB01
                line = line.replace("ﬂ", "fl")    # \uFB02
                line = line.replace("ﬃ  ", "ffi") # \uFB03
                line = line.replace("ﬄ  ", "ffl") # \uFB04
                pdftext.append('      ' + line)
            else:
                pdftext.append('    ' + line)
        pdftext.append('  </content>\n')
        if has_text:
            return pdftext
        else:
            return []

    except OSError as err:
        sys.stderr.write('(xdcrexiftool): pdftotext path (' + path + '): '
            + str(err) + '.\n')
        sys.exc_clear()
        return None 

class TransducerExifTool():
    '''
    Metadata extractor for various multimedia files.

        Starring the one and only, great and powerful
             *** ExifTool by Phil Harvey ***
        http://www.sno.phy.queensu.ca/~phil/exiftool/
    '''

    exiftool = None

    def __init__(self):
        '''Constructor'''
        self._init_exiftool()
        atexit.register(self._destroy_exiftool)

    def _init_exiftool(self):
        '''Create exiftool(1) process and keep it alive.'''
        try:
            self.exiftool = subprocess.Popen(
                [ 'exiftool', "-stay_open", "True", "-@", "-" ]
              , stdin=PIPE, stdout=PIPE, stderr=PIPE
            )

        except OSError as err:
            sys.stderr.write('(xdcrexiftool): Error ' + str(err) + '.\n')

    def _destroy_exiftool(self):
        '''Terminate exiftool(1) process.'''
        self.exiftool.stdin.write("-stay_open")
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.write("False")
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.flush()

    def expose_transducer(self):
        '''Report suffixes exiftool(1) can handle and a pointer to the
        handler aka extract function.'''
        # 215 suffixes in v8.60
        # $ for i in `exiftool -listf | grep -v Supp | \
        #   tr '[A-Z]' '[a-z]'`; do printf "'.$i', "; done;
        suffixes = [
            '.3fr', '.3g2', '.3gp', '.3gp2', '.3gpp', '.acfm', '.acr', '.afm',
            '.ai', '.aif', '.aifc', '.aiff', '.ait', '.amfm', '.ape', '.arw',
            '.asf', '.avi', '.bmp', '.btf', '.ciff', '.cos', '.cr2', '.crw',
            '.cs1', '.dc3', '.dcm', '.dcp', '.dcr', '.dfont', '.dib', '.dic',
            '.dicm', '.divx', '.djv', '.djvu', '.dll', '.dng', '.doc', '.docm',
            '.docx', '.dot', '.dotm', '.dotx', '.dv', '.dvb', '.dylib', '.eip',
            '.eps', '.eps2', '.eps3', '.epsf', '.erf', '.exe', '.exif', '.f4a',
            '.f4b', '.f4p', '.f4v', '.fla', '.flac', '.flv', '.fpx', '.gif',
            '.gz', '.gzip', '.hdp', '.htm', '.html', '.icc', '.icm', '.iiq',
            '.ind', '.indd', '.indt', '.itc', '.jng', '.jp2', '.jpeg', '.jpg',
            '.jpm', '.jpx', '.k25', '.kdc', '.key', '.kth', '.lnk', '.m2t',
            '.m2ts', '.m2v', '.m4a', '.m4b', '.m4p', '.m4v', '.mef', '.mie',
            '.mif', '.miff', '.mka', '.mks', '.mkv', '.mng', '.mos', '.mov',
            '.mp3', '.mp4', '.mpc', '.mpeg', '.mpg', '.mpo', '.mqv', '.mrw',
            '.mts', '.mxf', '.nef', '.newer', '.nmbtemplate', '.nrw',
            '.numbers', '.odp', '.ods', '.odt', '.ogg', '.orf', '.otf',
            '.pages', '.pbm', '.pct', '.pdf', '.pef', '.pfa', '.pfb', '.pfm',
            '.pgf', '.pgm', '.pict', '.pmp', '.png', '.pot', '.potm', '.potx',
            '.ppm', '.pps', '.ppsm', '.ppsx', '.ppt', '.pptm', '.pptx', '.ps',
            '.ps2', '.ps3', '.psb', '.psd', '.psp', '.pspframe', '.pspimage',
            '.pspshape', '.psptube', '.qif', '.qt', '.qti', '.qtif', '.ra',
            '.raf', '.ram', '.rar', '.raw', '.rif', '.riff', '.rm', '.rmvb',
            '.rpm', '.rsrc', '.rtf', '.rv', '.rw2', '.rwl', '.rwz', '.so',
            '.sr2', '.srf', '.srw', '.svg', '.swf', '.thm', '.thmx', '.tif',
            '.tiff', '.ts', '.ttc', '.ttf', '.tub', '.vob', '.vrd', '.vsd',
            '.wav', '.wdp', '.webm', '.webp', '.wma', '.wmv', '.x3f', '.xcf',
            '.xhtml', '.xla', '.xlam', '.xls', '.xlsb', '.xlsm', '.xlsx',
            '.xlt', '.xltm', '.xltx', '.xmp', '.zip'
        ]
        return (suffixes, self.extract_exiftool)

    # exiftool(1) is called with:
    #  $ exiftool -b -X --Exiftool:* --File:* --System:* <filename>
    #  0 <?xml version='1.0' encoding='UTF-8'?>
    #  1 <rdf:RDF xmlns:rdf='http://www.w3.org/1999/02/22-rdf-syntax-ns#'>
    #  2
    #  3 <rdf:Description rdf:about='/home/holu/Demo/Images/unisignet.jpg'
    #  4   xmlns:et='http://ns.exiftool.ca/1.0/' et:toolkit='Image::ExifTool 8.'
    #  5   xmlns:JFIF='http://ns.exiftool.ca/JFIF/JFIF/1.0/'
    #  6   xmlns:Ducky='http://ns.exiftool.ca/Ducky/Ducky/1.0/'
    #  7   xmlns:Adobe='http://ns.exiftool.ca/APP14/Adobe/1.0/'
    #  8   xmlns:Composite='http://ns.exiftool.ca/Composite/1.0/'>
    #  .  <JFIF:JFIFVersion>1.02</JFIF:JFIFVersion>
    #  .  <JFIF:ResolutionUnit>None</JFIF:ResolutionUnit>
    #  ...
    #  .  <FlashPix:UsedExtensionNumbers>
    #  .    <rdf:Bag>
    #  .      <rdf:li>1</rdf:li>
    #  .      <rdf:li>2</rdf:li>
    #  .    </rdf:Bag>
    #  .  </FlashPix:UsedExtensionNumbers>
    # -2 </rdf:Description>
    # -1 </rdf:RDF>
    #
    # Strategy:
    # - init output list with '<meta xmlns:rdf=...'
    # - skip first four lines from exif output (0-3)
    # - copy output (starting at 4) into output list
    # - minimal result (if no metadata is found):
    #   ['<meta xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#r"\n'
    #   , " xmlns:et='http://ns.exiftool.ca/1.0/' et:toolkit='...Tool 8.60'>\n"
    #   , '</rdf:Description>\n'
    #   , '</rdf:RDF>\n']
    # - do not return empty meta element, i.e. len(lines) == 4, but []
    # - otherwise delete last line -1 </rdf:RDF> from result
    #   and replace -2 </rdf:Description> with </meta>
    def extract_exiftool(self, path, suffix):
        '''Extracts metadata from path and constructs a result list.
        path  : string, absolute path to regular file
        return: list of strings. <meta>...</meta>'''

        self.exiftool.stdin.write("-b")
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.write("-X")
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.write("--Exiftool:*")
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.write("--File:*")
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.write("--System:*")
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.write(path)
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.write("-execute")
        self.exiftool.stdin.write("\n")
        self.exiftool.stdin.flush()

        line_count = 0
        lines = [
            '<meta xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#r"\n' ]
        while True:                                      # read from exiftool(1)
            line = self.exiftool.stdout.readline()
            if line.strip() == '{ready}':
                break
            if line_count < 4:                  # skip first lines from exiftool
                line_count += 1
                continue
            lines.append(' ' + line)
        if len(lines) <= 4:                          # no metadata from exiftool
            return []
        del lines[-2:] # has metadata, remove '</rdf:RDF>', '</rdf:Description>'
        if suffix == '.pdf':
            fulltext = extract_pdftohtml(path)
            if fulltext:
                lines.extend(fulltext)
        lines.append('</meta>\n')      # replace </rdf:Description> with </meta>
        return lines
